File: CRO/mCRO/RO_solver_analytic.m
function [T_anal_out,h_anal_out,noise_out]=RO_solver_analytic(par,IC,N,NE,dt,saveat,savemethod,noise_custom)
%% Solve Analytical Solution for Linear White RO 
% White Noise RO (ntype==1)
% dT/dt=R*T+F1*h+(sigma_T)*(xi_T)
% dh/dt=-F2*T-epsilon*h+(sigma_h)*(xi_h)
% where xi_T and xi_h are white noise with zero mean and unit variance

%% Setting Default Option
if nargin<5||isempty(dt)
    dt=0.1; 
end
if nargin<6||isempty(saveat)
    saveat=1.0;  
end
if nargin<7||isempty(savemethod)
    savemethod="sampling";
end
if nargin<8||isempty(noise_custom)
    noise_custom=[];  
end

%% Setting Time
NT=round(N/dt);

%% Setting Noise
if isempty(noise_custom)
    noise_custom=randn(NT-1,4,NE);
else
    if isscalar(noise_custom)
        rng(noise_custom)
        noise_custom=randn(NT-1,4);
        noise_custom=repmat(noise_custom,[1 1 NE]); % note: must keep consistency with noise generator in RO_solver        
    end 
end 

%% Setting Parameter
% post-processing par
for i=1:length(par)
    par_sel(i)=par{i}(1); % only select annual mean value
end 

% linear parameter
R=par_sel(1);
F1=par_sel(2);
F2=par_sel(3);
epsilon=par_sel(4);

% noise parameter
sigma_T=par_sel(9);
sigma_h=par_sel(10);

%% Setting Values 
To=IC(1);
ho=IC(2);

t=0:dt:(NT-1)*dt;
t=transpose(t); 

gr=(R-epsilon)/2;
w=sqrt(4*F1*F2-(R+epsilon)^2)/2;

%% Calculate Deterministic Part
T_anal_det=exp(gr*t).*(To*cos(w*t)+To*(R+epsilon)/(2*w)*sin(w*t)+ho*(F1/w)*sin(w*t));
h_anal_det=exp(gr*t).*(ho*cos(w*t)-ho*(R+epsilon)/(2*w)*sin(w*t)-To*(F2/w)*sin(w*t));

%% Calculate Stochastic Part
for j=1:NE
w_T=sigma_T*squeeze(noise_custom(:,1,j))/sqrt(dt);
w_h=sigma_h*squeeze(noise_custom(:,2,j))/sqrt(dt);

T_anal_sto=zeros(length(t),1);
h_anal_sto=zeros(length(t),1);

for i=1:length(t)
    if i==1
        T_anal_sto(1,1)=0.0;
        h_anal_sto(1,1)=0.0;
    else
        tau=t(1:i-1);
        T_anal_sto(i,1)=sum(dt*exp(gr*(t(i)-tau)).*(w_T(1:i-1).*cos(w*(t(i)-tau))+w_T(1:i-1).*(R+epsilon)/(2*w).*sin(w*(t(i)-tau))+w_h(1:i-1).*(F1/w).*sin(w*(t(i)-tau))),1);
        h_anal_sto(i,1)=sum(dt*exp(gr*(t(i)-tau)).*(w_h(1:i-1).*cos(w*(t(i)-tau))-w_h(1:i-1).*(R+epsilon)/(2*w).*sin(w*(t(i)-tau))-w_T(1:i-1).*(F2/w).*sin(w*(t(i)-tau))),1);
    end
end 

%% Total Solution 
T_anal=T_anal_det+T_anal_sto;
h_anal=h_anal_det+h_anal_sto;

T_anal_out(:,j)=T_anal;
h_anal_out(:,j)=h_anal;
end 

%% Select Time Step % Save Noise Output
% note: noise_out does not follow this savemethod option 
if savemethod=="sampling"
    T_anal_out=T_anal_out(1:round(saveat/dt):end,:);
    h_anal_out=h_anal_out(1:round(saveat/dt):end,:);
elseif savemethod=="mean"
    T_out_sel=T_anal_out(0.5*round(1/dt):end-0.5*round(1/dt)-1,:);
    h_out_sel=T_anal_out(0.5*round(1/dt):end-0.5*round(1/dt)-1,:);  
    T_out_sel=reshape(T_out_sel,round(saveat/dt),[],size(T_out_sel,2));
    T_out_sel=squeeze(mean(T_out_sel,1));
    h_out_sel=reshape(h_out_sel,round(saveat/dt),[],size(h_out_sel,2));
    h_out_sel=squeeze(mean(h_out_sel,1));
    if NE>1
    T_anal_out=[T_anal_out(1,:);T_out_sel];
    h_anal_out=[h_anal_out(1,:);h_out_sel];
    elseif NE==1
    T_anal_out=[T_anal_out(1),T_out_sel];
    h_anal_out=[h_anal_out(1),h_out_sel];
    T_anal_out=transpose(T_anal_out);
    h_anal_out=transpose(h_anal_out);    
    end 
end
noise_out=noise_custom; 

end 

File: CRO/pyCRO/RO_solver_analytic.py
import numpy as np
import sys
import matplotlib.pyplot as plt

def RO_solver_analytic(par, IC, N, NE, dt=0.1, saveat=1.0, savemethod="sampling", noise_custom=None):
    NT = int(round(N / dt))
    step = int(round(saveat / dt))
    ratio = saveat / dt

    ### Checking simulation setups ###
    ##################################
    print("---------------------------------------------------------------------------------")
    print("Welcome to the CRO Analytical Solver!")
    print("Ensure that the same arguments are provided as for RO_solver,\n"
          "except that 'NM' and 'EF' are not required.")
    print("---------------------------------------------------------------------------------")

    ### Check timestep settings ###
    if dt > saveat:
        raise ValueError(f"dt={dt} and saveat={saveat}: The numerical time step (dt) must be smaller than or equal to the data saving interval (saveat).")
    if not ratio.is_integer():
        raise ValueError(f"dt = {dt}, saveat = {saveat}: saveat must be divisible by dt so that saveat/dt results in an integer.")
    ###############################

    ### extracting linear and noise amplitude parameters ###
    ########################################################
    R_value = par['R'][0]
    F1_value = par['F1'][0]
    epsilon_value = par['epsilon'][0]
    F2_value = par['F2'][0]
    sigma_T_value = par['sigma_T'][0]
    sigma_h_value = par['sigma_h'][0]
    ########################################################
    ########################################################


    ### initial setups ###
    ######################
    To, ho = IC
    t = np.arange(NT) * dt
    gr = (R_value - epsilon_value) / 2
    w = np.sqrt(4 * F1_value * F2_value - (R_value + epsilon_value)**2) / 2

    T_anal_out = np.zeros((NT, NE))
    h_anal_out = np.zeros((NT, NE))

    T_det = np.exp(gr * t) * (To * np.cos(w * t) +
        To * (R_value + epsilon_value) / (2 * w) * np.sin(w * t) +
        ho * (F1_value / w) * np.sin(w * t))

    h_det = np.exp(gr * t) * (ho * np.cos(w * t) -
        ho * (R_value + epsilon_value) / (2 * w) * np.sin(w * t) -
        To * (F2_value / w) * np.sin(w * t))
    ######################
    ######################


    ### analytical calculations for each ensemble member ###
    ########################################################
    noise_out = []
    for j in range(NE):
        if noise_custom is None:
            noise_array = np.random.randn(NT - 1, 4) # It generates noises with shapes (NT, 4, NE)
        elif np.isscalar(noise_custom):
            np.random.seed(noise_custom)             # It generates noises with seed number = noise_custom
            noise_array = np.random.randn(NT - 1, 4) # Note that noises are same for all the ensemble members
        elif noise_custom.shape == np.random.randn(NT - 1, 4).shape:
            noise_array = noise_custom               # It uses user-provided noises
        else:
            raise ValueError(f"Invalid 'noise_custom' input: expected 'None', a scalar (int or float), "
                             f"or an array with shape (N/dt-1, 4) = ({NT-1}, 4).")
        noise_out.append(noise_array)
        w_T = sigma_T_value * noise_array[:, 0] / np.sqrt(dt)
        w_h = sigma_h_value * noise_array[:, 1] / np.sqrt(dt)

        T_sto = np.zeros(NT) 
        h_sto = np.zeros(NT) 

        for i in range(1, NT): 
            tau = t[:i] 
            delta = t[i] - tau
            sin = np.sin(w * delta)
            cos = np.cos(w * delta)

            T_sto[i] = np.sum(dt * np.exp(gr * delta) * (
                w_T[:i] * cos +                                        
                w_T[:i] * (R_value + epsilon_value) / (2 * w) * sin +
                w_h[:i] * (F1_value / w) * sin
            ))

            h_sto[i] = np.sum(dt * np.exp(gr * delta) * (
                w_h[:i] * cos -                                        
                w_h[:i] * (R_value + epsilon_value) / (2 * w) * sin -
                w_T[:i] * (F2_value / w) * sin
            ))

        T_anal = T_det + T_sto 
        h_anal = h_det + h_sto 

        T_anal_out[:, j] = T_anal 
        h_anal_out[:, j] = h_anal 
    noise_out = np.array(noise_out)
    noise_out = np.transpose(noise_out, (1, 2, 0))
    ########################################################
    ########################################################


    ### save data ###
    #################
    if savemethod == "sampling":
        T_anal_out = T_anal_out[::step, :]
        h_anal_out = h_anal_out[::step, :]
    elif savemethod == "mean":
        T_anal_out = T_anal_out[int(np.ceil(step/2)):NT-int(step/2),:]
        h_anal_out = h_anal_out[int(np.ceil(step/2)):NT-int(step/2),:]
        T_anal_out = T_anal_out.reshape(-1, step, NE).mean(axis=1)
        h_anal_out = h_anal_out.reshape(-1, step, NE).mean(axis=1)
        if NE > 1:
            T_anal_out = np.vstack([T_anal_out[0:1, :], T_anal_out])
            h_anal_out = np.vstack([h_anal_out[0:1, :], h_anal_out])
        elif NE == 1:
            T_anal_out = np.concatenate([[T_anal_out[0]], T_anal_out])
            h_anal_out = np.concatenate([[h_anal_out[0]], h_anal_out])
    print("All steps are successfully completed!")
    print("---------------------------------------------------------------------------------")
    #################
    #################
    return T_anal_out, h_anal_out, noise_out


